subsys/net/ip/net_tls.c

```diff
@@ -111,3 +126,42 @@ int net_tls_connect(struct net_context *context, bool listening)
 
 	return err;
 }
+
+int net_tls_send(struct net_pkt *pkt)
+{
+	struct net_context *context = net_pkt_context(pkt);
+	struct net_buf *frag = pkt->frags;
+
+	while (frag) {
+		int ret = 0, tries = 3;
+		u8_t *data = frag->data;
+		u16_t len = frag->len;
+
+		while (ret < len) {
+			ret = mbedtls_ssl_write(&context->mbedtls.ssl,
```

Comment:
Sending user input data over TLS fragment by fragment is not ideal (adds overhead, limits throughput). Zstream API is free from such a stipulation (it actually has API-level support for arbitrary buffering, if useful to the underlying implementation).

Reply:
We need to figure out how to use mbedtls, I was hoping it were good enough in grouping fragmented writes. If not, I need to look at Zstream code, which I was thinking of using provided that I first understand what mbedtls is in general doing.

Reply:
Here I had to demonstrate that no bytes get left behind with suboptimal mbedtls usage, or make everybody notice that not all fragments were accounted for. mbedtls API is not that nice, nor does it understand Zephyr buffers and packets. But that is business as usual.